### ISSUE TITLE
A more robust way to deprecate APIs

### ISSUE THREAD

**Issue by ndrluis** [apache/iceberg-python#1330]
### Feature Request / Improvement

I was studying Python libraries and how they handle deprecation. Specifically, I was exploring constant deprecation in the context of issue #1217. I found this [module](https://github.com/conda/conda/blob/main/conda/deprecations.py) in Conda, and they have [good documentation about deprecation](https://docs.conda.io/projects/conda/en/latest/dev-guide/deprecations.html). Since BSD-3 is compatible with Apache 2.0, I believe we could adopt their approach to deprecation. It is more robust and provides the ability to deprecate constants in a module or arguments in a function.

I’d love to hear what you all think about this. Are there any suggestions or concerns we should consider?

**Comment by kevinjqliu** [comment]
thanks for doing the research! I like this approach. 
do you know if the deprecation message includes the call site or stack trace? for example, in #1336, it would be helpful to know which function produced the deprecation message

**Comment by Fokko** [comment]
Hey @ndrluis thanks for bringing this up! Are you suggesting to copy the code into our codebase? I always favor reusing an existing library instead of reinventing the wheel. I also noticed that they refer to [deprecation](https://deprecation.readthedocs.io/en/latest/) in the code, which might also be a good candidate.

**Comment by ndrluis** [comment]
@kevinjqliu Yes, it refers to the call site. The ‘argparse.Action’ example in the documentation demonstrates this.

@Fokko I agree with using a library, but the deprecation library has fewer features than conda.deprecation. One of the most important features I am looking for is the constant deprecation that exists in Conda. I think that, in our case, it makes more sense to copy the deprecation source code from the Conda codebase rather than having Conda as a dependency.

**Comment by Fokko** [comment]
My concern is that I don't see `conda` on `pypi`, and we don't want to force users to `condaforge`.

If the license is compatible, we can also copy the code, and mention it in the `LICENSE`: https://github.com/apache/iceberg-python/blob/7a83695330518bea0dee589b5b513297c4d59b66/LICENSE#L225

**Comment by ndrluis** [comment]
One new aspect introduced with this Conda deprecation code is the concept of a pending deprecation process. I think that when we release version 1.0, we will need to allocate more time to remove deprecated code. The idea is to establish a timeline for pending deprecations, during which users can dispute the decision to deprecate a specific API. I believe this is a valuable feature, as we already have a relevant example in issue #1104.

More details about the Conda deprecation policy are available here: https://github.com/conda/ceps/blob/main/cep-0009.md.

I plan to add a section to our documentation about the deprecation policy, following the guidelines outlined here: https://iceberg.apache.org/contribute/?h=deprecation#major-version-deprecations-required.

What do you all think?
